modules/auxiliary/scanner/msmail/exchange_enumerator.go

```diff
@@ -0,0 +1,543 @@
+//usr/bin/env go run "$0" "$@"; exit "$?"
+
+/*
+ OWA (Outlook Webapp) is vulnerable to time-based user enumeration attacks.
+ This module leverages all known, and even some lesser-known services exposed by default
+ Exchange installations to enumerate users. It also targets Office 365 for error-based user enumeration.
+
+ Identify Command
+
+    Used for gathering information about a host that may be pointed towards an Exchange or o365 tied domain
+    Queries for specific DNS records related to Office 365 integration
+    Attempts to extract internal domain name for onprem instance of Exchange
+    Identifies services vulnerable to time-based user enumeration for onprem Exchange
+    Lists password-sprayable services exposed for onprem Exchange host
+
+ Userenum (o365) Command
+
+    Error-based user enumeration for Office 365 integrated email addresses
+
+ */
+package main
+
+import (
+	"crypto/tls"
+	b64 "encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"metasploit/module"
+	"net"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+	"strconv"
+)
+
+func main() {
+	metadata := &module.Metadata{
+		Name:        "msmailprobe",
+		Description: "Office 365 and Exchange Enumeration",
+		Authors:     []string{"poptart", "jlarose", "Vincent Yui", "grimhacker", "Nate Power", "Nick Powers", "clee-r7"},
+		Date:        "2018-11-6",
+		Type:        "single_scanner",
+		Privileged:  false,
+		References:  []module.Reference{},
+		Options: map[string]module.Option{
+			"Command":          {Type: "string", Description: "Either 'userenum' or 'identify'", Required: true, Default: "identify"},
+			"OnPrem":           {Type: "bool", Description: "Flag to specify an On-Premise instance of Exchange", Required: false, Default: "false"},
+			"O365":             {Type: "bool", Description: "Use this flag if Exchange services are hosted by Office 365", Required: false, Default: "false"},
+			"UserName":         {Type: "string", Description: "Single user name to do identity test against", Required: false, Default: ""},
+			"UserNameFilePath": {Type: "string", Description: "Path to file containing list of users", Required: false, Default: ""},
+			"Email":            {Type: "string", Description: "Single email address to do identity test against", Required: false, Default: ""},
+			"EmailFilePath":    {Type: "string", Description: "Path to file containing list of email addresses", Required: false, Default: ""},
+			"OutputFile":       {Type: "string", Description: "Used for outputting valid users/email", Required: false, Default: ""},
```

Comment:
this should be using the creds API for Metsploit rather than writing to a file. You want the module to report creds like any other module.

Reply:
This does not output credentials.

Reply:
Here are some tutorials on the wiki about how login scanners work:

https://github.com/rapid7/metasploit-framework/wiki/Creating-Metasploit-Framework-LoginScanners

https://github.com/rapid7/metasploit-framework/wiki/How-to-write-a-HTTP-LoginScanner-Module

you're forging some new ground since this is an external module, and in a new language, so some similar facilities that interact with Metasploit's core will have to be created to make this module work like others.